In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", 50)
print("Libraries loaded sucessfully")

Libraries loaded sucessfully


In [2]:
#load dataset
df = pd.read_csv("../data/diabetic_data.csv")
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

Rows: 101766
Columns: 50


In [9]:
#Preview data
df.head(10)

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,payer_code,medical_specialty,num_lab_procedures,num_procedures,num_medications,number_outpatient,number_emergency,number_inpatient,diag_1,diag_2,diag_3,number_diagnoses,max_glu_serum,A1Cresult,metformin,repaglinide,nateglinide,chlorpropamide,glimepiride,acetohexamide,glipizide,glyburide,tolbutamide,pioglitazone,rosiglitazone,acarbose,miglitol,troglitazone,tolazamide,examide,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,?,Pediatrics-Endocrinology,41,0,1,0,0,0,250.83,?,?,1,NaN,NaN,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,?,?,59,0,18,0,0,0,276,250.01,255,9,NaN,NaN,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,?,?,11,5,13,2,0,1,648,250,V27,6,NaN,NaN,No,No,No,No,No,No,Steady,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,?,?,44,1,16,0,0,0,8,250.43,403,7,NaN,NaN,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,?,?,51,0,8,0,0,0,197,157,250,5,NaN,NaN,No,No,No,No,No,No,Steady,No,No,No,No,No,No,No,No,No,No,Steady,No,No,No,No,No,Ch,Yes,NO
5,35754,82637451,Caucasian,Male,[50-60),?,2,1,2,3,?,?,31,6,16,0,0,0,414,411,250,9,NaN,NaN,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,Steady,No,No,No,No,No,No,Yes,>30
6,55842,84259809,Caucasian,Male,[60-70),?,3,1,2,4,?,?,70,1,21,0,0,0,414,411,V45,7,NaN,NaN,Steady,No,No,No,Steady,No,No,No,No,No,No,No,No,No,No,No,No,Steady,No,No,No,No,No,Ch,Yes,NO
7,63768,114882984,Caucasian,Male,[70-80),?,1,1,7,5,?,?,73,0,12,0,0,0,428,492,250,8,NaN,NaN,No,No,No,No,No,No,No,Steady,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,No,Yes,>30
8,12522,48330783,Caucasian,Female,[80-90),?,2,1,4,13,?,?,68,2,28,0,0,0,398,427,38,8,NaN,NaN,No,No,No,No,No,No,Steady,No,No,No,No,No,No,No,No,No,No,Steady,No,No,No,No,No,Ch,Yes,NO
9,15738,63555939,Caucasian,Female,[90-100),?,3,3,4,12,?,InternalMedicine,33,3,18,0,0,0,434,198,486,8,NaN,NaN,No,No,No,No,No,No,No,No,No,No,Steady,No,No,No,No,No,No,Steady,No,No,No,No,No,Ch,Yes,NO


In [3]:
#check datatype
df.dtypes

encounter_id                 int64
patient_nbr                  int64
race                        object
gender                      object
age                         object
weight                      object
admission_type_id            int64
discharge_disposition_id     int64
admission_source_id          int64
time_in_hospital             int64
payer_code                  object
medical_specialty           object
num_lab_procedures           int64
num_procedures               int64
num_medications              int64
number_outpatient            int64
number_emergency             int64
number_inpatient             int64
diag_1                      object
diag_2                      object
diag_3                      object
number_diagnoses             int64
max_glu_serum               object
A1Cresult                   object
metformin                   object
repaglinide                 object
nateglinide                 object
chlorpropamide              object
glimepiride         

In [4]:
# Find missing values (the ? problem)
#replace all "?", with NaN first
df.replace("?", np.nan, inplace=True)

#Now count missing values per column
missing = df.isnull().sum()
missing_pct = (missing / len(df) *100).round(1)

# create a summary table of missing values
missing_summary = pd.DataFrame({
    "missing_count" : missing,
    "missing_pct"   : missing_pct
}).query("missing_count > 0").sort_values( "missing_pct", ascending=False )

print(missing_summary )

                   missing_count  missing_pct
weight                     98569         96.9
max_glu_serum              96420         94.7
A1Cresult                  84748         83.3
medical_specialty          49949         49.1
payer_code                 40256         39.6
race                        2273          2.2
diag_3                      1423          1.4
diag_2                       358          0.4
diag_1                        21          0.0


In [5]:
#Understanding target variable
target_counts = df["readmitted"].value_counts()
target_pct   = df["readmitted"].value_counts(normalize=True).mul(100).round(1)

#create a summary table for target variable
target_summary = pd.DataFrame({
    "Count"       : target_counts,
    "Percentage"  : target_pct
    })
print(target_summary)

            Count  Percentage
readmitted                   
NO          54864        53.9
>30         35545        34.9
<30         11357        11.2


In [13]:
#cleaning all the missing values
#Drop columns with too much missing data
cols_to_drop = ["weight", "max_glu_serum", "medical_specialty", "payer_code"]
df.drop(columns=cols_to_drop, inplace=True)
print(f"Columns remaining: {df.shape[1]}") 

Columns remaining: 46


In [14]:
#Handling A1Cresult(turn missing into feature)
#Missing A1Cresult means the test was never ordered 
#This is actually a meaningful clinical info-create a binary feature 
df["A1C_tested"] = df["A1Cresult"].apply(lambda x: 0 if pd.isna(x) else 1)


#For the original column, fill missing with "Not tested"
df["A1Cresult"] = df["A1Cresult"].fillna("Not test")

print(df["A1C_tested"].value_counts())
print(df["A1Cresult"].value_counts())
                                

A1C_tested
0    84748
1    17018
Name: count, dtype: int64
A1Cresult
Not test    84748
>8           8216
Norm         4990
>7           3812
Name: count, dtype: int64


In [6]:
#Handling race and diagnosis column
#Fill missing race with most frequent value
df["race"] = df["race"].fillna(df["race"].mode()[0])

#Drop the small number of rows missing diagnosis codes
df.dropna(subset=["diag_1", "diag_2", "diag_3"], inplace=True)

print(f"Rows after  dropping missing diagnoses: {len(df)}")

Rows after  dropping missing diagnoses: 100244


In [7]:
#Handling duplicate patients
#Dataset has multiple visits per patient-keep only first encounter
#This prevents data leakage(a patient's later visits infleuncing earlier predictions)
df = df.sort_values("time_in_hospital", ascending=False)
df = df.drop_duplicates(subset="patient_nbr", keep="first")

print(f"Rows after removing duplicate patients: {len(df)}")

Rows after removing duplicate patients: 70416


In [8]:
#Creating he binary target variable
#Convert the readmitted to binary : 1 = readmitted within 30 days, 0 = everytthing else
df["readmitted_30"] = df["readmitted"].apply(lambda x: 1 if x == "<30" else 0)

# check the class balance
counts = df["readmitted_30"].value_counts()
pcts   = df["readmitted_30"].value_counts(normalize=True).mul(100).round(1)

print(pd.DataFrame({"counts" : counts, "percentage" : pcts}))

               counts  percentage
readmitted_30                    
0               65582        93.1
1                4834         6.9


In [9]:
# confirming the claen dataset
print(f"Final dataset shape: {df.shape}")
print(f"\nAny remaining missing values?")
print(df.isnull().sum()[df.isnull().sum()>0])

#Save the clean data version for use in the next phase
df.to_csv("../data/diabetic_data_clean.csv", index=False)

Final dataset shape: (70416, 51)

Any remaining missing values?
weight               67578
payer_code           29906
medical_specialty    34203
max_glu_serum        66976
A1Cresult            57766
dtype: int64


**Data summary**

65,582 patients not readmitted within 30 days

4,834 patients readmitted within 30 days

The minority class is roughly 1 in 14 patients


In [10]:
#Encode the target and remove irrevant ccolumns
#Drop columns that are IDs or leak feature informations
cols_to_remove = ["encounter_id", "patient_nbr", "readmitted"]
df.drop(columns=cols_to_remove, inplace=True)

print(f"Shape: {df.shape}")

Shape: (70416, 48)


In [11]:
#Convert age brackets to numbers
#Age comes as brackets like [50-60), convert to ordinal numbers
age_map = {
    "[0-10)":1, "[10-20)":2,  "[20-30)" :3, "[30-40)" :4, "[40-50)":5, 
    "[50-60)":6, "[60-70)":7, "[70-80)":8, "[80-90)":9, "[90-100)":1
}

df["age_num"] = df["age"].map(age_map)
df.drop(columns=["age"], inplace=True)

print(df["age_num"].value_counts().sort_index())


age_num
1     2029
2      353
3     1029
4     2562
5     6646
6    12266
7    15795
8    18069
9    11667
Name: count, dtype: int64


In [40]:
#Creating clinical risk features

#Total procedures and medications as a complexity proxy
df["total_interactions"] = df["num_medications"] + df["num_procedures"] + df["num_lab_procedures"]

#Prior hospital usage -strong readmission predictions
df["prior_hospital_use"] = df["number_outpatient"] + df["number_emergency"] + df["number_inpatient"]

#flag long stays (7+ days often indcates severity)
df["long_stay"] = (df["time_in_hospital"] >=7).astype(int)

print(df[["total_interactions", "prior_hospital_use", "long_stay"]].describe())


       total_interactions  prior_hospital_use     long_stay
count        70416.000000        70416.000000  70416.000000
mean            62.259004            0.779397      0.253479
std             24.246330            1.698018      0.435006
min              2.000000            0.000000      0.000000
25%             47.000000            0.000000      0.000000
50%             62.000000            0.000000      0.000000
75%             78.000000            1.000000      1.000000
max            179.000000           68.000000      1.000000


In [12]:
#simplifying 23 medication columns
#Each drug column has values No, steady, Up, Down
#Encode: 0=not prescribed, 1 = prescribed(any dose)
med_cols = [
    "metformin", "repaglinide", "nateglinide", "chlorpropamide",
            "glimepiride", "acetohexamide", "glipizide", "glyburide",
            "tolbutamide", "pioglitazone", "rosiglitazone", "acarbose",
            "miglitol", "troglitazone", "tolazamide", "examide",
            "citoglipton", "insulin", "glyburide-metformin",
            "glipizide-metformin", "glimepiride-pioglitazone",
            "metformin-rosiglitazone", "metformin-pioglitazone"
]

for col in med_cols:
    df[col] = df[col].apply(lambda x : 0 if  x == "No" else 1)

#Also create a total meds count feature
df["total_meds_prescribed"] = df[med_cols].sum(axis=1)

print(f"Avg medications prescribed: {df["total_meds_prescribed"].mean():.1f}")

Avg medications prescribed: 1.2


In [13]:
#Encode remaining categorical columns
from sklearn.preprocessing import LabelEncoder

cat_cols = ["race", "gender", "A1Cresult", "change", "diabetesMed"]

le = LabelEncoder()
for col in cat_cols:
    df[col] = le.fit_transform(df[col].astype(str))

print("Categorical columns encoded.")
print(df[cat_cols].head())

Categorical columns encoded.
       race  gender  A1Cresult  change  diabetesMed
39487     2       1          3       1            1
39500     2       0          3       1            0
88650     0       1          2       0            1
59150     0       0          3       1            0
59156     2       1          3       1            1


In [14]:
#droping diagnosis columns for now
#diag_1/2/3 are ICD _9 codes with 700+ unique values
#Grouping them properly takes a fulll notebook on its own
#For v1 of this project, droping this the best option now

df.drop(columns=["diag_1", "diag_2", "diag_3"], inplace=True)

print(f"Final shape: {df.shape}")
print(f"\nAll columns:\n{df.columns}")

Final shape: (70416, 46)

All columns:
Index(['race', 'gender', 'weight', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'payer_code', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'number_diagnoses',
       'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted_30',
       'age_num', 'total_meds_prescribed'],
      dtype='object')


In [49]:
# Check where the typo came from
print(df.columns[df.columns.str.contains('meds')].tolist())

# Drop the typo column
df.drop(columns=['total_meds_ prescribed'], inplace=True)

# Confirm it's gone
print(f"Fixed. Shape: {df.shape}")
print(df.columns.tolist())

['total_meds_ prescribed', 'total_meds_prescribed']
Fixed. Shape: (70416, 46)
['race', 'gender', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'A1C_tested', 'readmitted_30', 'age_num', 'total_interactions', 'prior_hospital_use', 'long_stay', 'total_meds_prescribed']


In [15]:
#Saving and do final checks
#confirm no missing values remain
print("Missing values remaining:")
print(df.isnull().sum()[df.isnull().sum()>0])

# save for modelling notebook
df.to_csv("../data/diabetic_data_features.csv", index=False)
print(f"\nFeature dataset saved: shape: {df.shape}")

Missing values remaining:
weight               67578
payer_code           29906
medical_specialty    34203
max_glu_serum        66976
dtype: int64

Feature dataset saved: shape: (70416, 46)
